### PR TITLE
Add ESLint import resolver

### PR DESCRIPTION
**Pull request by kevin940726**
## Description

[`eslint-plugin-import`](https://github.com/benmosher/eslint-plugin-import) is using the default [import resolver](https://github.com/benmosher/eslint-plugin-import#resolvers) to resolve the packages. By default, it's looking for the "main" field in `package.json`. This works for most packages, but not automatically for internal `@wordpress/*` packages. We need to at least run `npm run dev` or `npm run build` once to generate the build files in order for the resolver to work. It isn't a big deal though, but wouldn't it be better if we can skip this step and just automatically reference the file to their source?

This PR introduces a new internal package `@wordpress/eslint-import-resolver` to handle the resolution of the `@wordpress/*` packages. It tries to fallback to finding the source file in `src/index.js` whenever the "main" file isn't present. For other external dependencies, it simply fallbacks to the default resolution using `eslint-import-resolver-node`.

The benefit of this is that it no longer requires the developers to build the packages before they can run `lint-js`, or before the code editor's built-in linter shout at them. 

## How has this been tested?

Return to a "clean state" of the project, that is, remove all of the `build` folders.

```sh
rm -rf packages/**/build packages/**/build-module packages/**/build-types packages/**/tsconfig.tsbuildinfo
```

Install the dependencies, and then `lint-js` script should work without extra steps. Or, do a clean clone of the project to a new directory.

```sh
npm install
npm run lint-js
```

## Types of changes

New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: .github/workflows/static-checks.yml [hunk 1]
```diff
@@ -31,22 +31,20 @@ jobs:
...
-                  npm run build
+              run: npm install
 
             - name: Lint JavaScript and Styles
               run: npm run lint
 
-            - name: Type checking
-              run: npm run build:package-types
+            - name: Build and type checking
+              run: npm run build
```
Comment (kevin940726) [inline .github/workflows/static-checks.yml]: We probably don't even need to run the full build in this workflow? Just `build:package-types` is enough?

**Comment by gziolo** [comment]
> We need to at least run `npm run dev` or `npm run build` once to generate the build files in order for the resolver to work

Great idea to use the custom resolver to remove the need for the build step for Gutenberg. We do something similar already for unit tests with a Jest mapper 👍🏻 

> This PR introduces a new internal package @wordpress/eslint-import-resolver to handle the resolution of the @wordpress/* packages.

I don't think we need a package. It isn't applicable outside of Gutenberg. It will make this PR simpler.

**Comment by kevin940726** [comment]
> I don't think we need a package. It isn't applicable outside of Gutenberg

Sure, but where should I put this file then? I don't think it belongs in `eslint-plugin` as it's only useful in this project's scope.

**Comment by gziolo** [comment]
> Sure, but where should I put this file then?

There is the `test` folder, so you can try:

`test/lint/import-resolver.js`

File: .github/workflows/static-checks.yml (hunk 1 above)
Comment (kevin940726) [reply .github/workflows/static-checks.yml]: I removed the build step. We can add it back if we really need it.

File: test/lint/import-resolver.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
+/**
+ * External dependencies
+ */
+const resolverNode = require( 'eslint-import-resolver-node' );
```
Comment (gziolo) [inline test/lint/import-resolver.js:4]: `eslint-import-resolver-node` should be listed as a dev dependency in the main `package.json` file.

File: test/lint/import-resolver.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+exports.resolve = function ( source, file, config ) {
+	if ( source.startsWith( '@wordpress/' ) ) {
+		const packageName = source.slice( '@wordpress/'.length );
+
+		let entryFile;
+		try {
+			entryFile = require.resolve( `../../packages/${ packageName }` );
+		} catch {
+			entryFile = require.resolve(
+				`../../packages/${ packageName }/src/index.js`
```
Comment (gziolo) [inline test/lint/import-resolver.js]: Can it omit the file name? I theory, the entry point could be now `index.ts` or even less likely but possible `index.tsx`.

**Review by gziolo — APPROVED** [review]
Nice, thank you for working on this improvement for CI.

I left two minor comments to address before merging.

**Comment by gziolo** [comment]
@youknowriad noticed that WordPress dependencies can be added to the code and ESLint doesn't report it as a violation. It was caught during backporting to the `wp/trunk` branch that doesn't contain changes from this PR see https://github.com/WordPress/gutenberg/pull/32848#issuecomment-865284371:

> There's a lint failure on this PR. It's not clear why it's only failing on this branch and not in trunk and it's being fixed in #32866

Those violations were already fixed in the `trunk`, but we should look into improvements to this implementation so it not only checks if the package is present but more importantly if it is listed as a package dependency.

The simplest way to test is to disable the resolver and import anything from `@wordpress/editor` in the `@wordpress/block-library` package as it was the case in `trunk`.

**Comment by kevin940726** [comment]
Good catch! I indeed missed this case, I'll see what I can do in a follow-up PR.

**Pull request merged**